docs/proposals/2841-flux-hpc/README.md

```diff
@@ -0,0 +1,275 @@
+
+```go
+type MLPolicySource struct {
+    Torch *TorchMLPolicySource `json:"torch,omitempty"`
+    MPI   *MPIMLPolicySource   `json:"mpi,omitempty"`
+    HPC   *HPCMLPolicySource   `json:"hpc,omitempty"`
+
+    // FluxMLPolicy defines policy only for Flux
+    Flux  *FluxMLPolicySource  `json:"flux,omitempty"`
+}
+
+
+// FluxMLPolicySource represents a Flux-specific runtime configuration 
+// Fields must be hard-coded fields.
+type FluxMLPolicySource struct {
+    // Tasks is the number of tasks for the Flux job.
+    // +optional
+    Tasks *int32 `json:"tasks,omitempty"`
+
+    // ViewImage specifies the container image for the Flux init container.
+    // +optional
+    ViewImage string `json:"viewImage,omitempty"`
+
+    // NetworkDevice is the network interface Flux should use for communication (e.g., "eth0").
+    // +optional
+    NetworkDevice string `json:"networkDevice,omitempty"`
+
+    // QueuePolicy sets the scheduling policy for the Fluxion scheduler (e.g., "fcfs").
+    // +kubebuilder:validation:Enum=fcfs;easy
+    // +optional
+    QueuePolicy string `json:"queuePolicy,omitempty"`
+
+    // Interactive specifies whether to run an interactive Flux shell.
+    // +optional
+    Interactive *bool `json:"interactive,omitempty"`
+}
+```
+
+There would likely be many more fields. The above is a subset as an example that we have implemented so far.
+
+##### Example `ClusterTrainingRuntime` with `fluxPolicy`
+
+```yaml
+apiVersion: trainer.kubeflow.org/v1alpha1
+kind: ClusterTrainingRuntime
+metadata:
+  name: hpc-flux-specific-runtime
+spec:
+  mlPolicy:
+    flux:
+      tasks: 4
+      viewImage: "ghcr.io/converged-computing/flux-view-ubuntu:tag-jammy"
+      networkDevice: "eth0"
+      queuePolicy: "fcfs"
+      interactive: false
+```
+
+While this approach offers the benefits of strong typing and explicit, self-documenting fields, there are several risks to be mitigated:
+
+1.  **Tight Coupling and API Brittleness:** This design tightly couples the Kubeflow Trainer API to the specific implementation details of Flux Framework. If Flux itself introduces a new, important configuration option, it would require a formal change to the Kubeflow Trainer API. E.g., if we make additional options to Flux a string field, that serves as a mitigation for API brittleness. Requiring each flag or option to be a hardened field would make future development harder and require a new release. This creates a high maintenance burden and slows down the adoption of new features.
+
+2.  **Lack of Extensibility:** The primary risk is inflexibility. The `hpcPolicy` with a generic `manager` and `settings` map was chosen because it can support other HPC workload managers (like Slurm, LSF, PBS) in the future *without any changes to the API*. A hard-coded approach would necessitate adding a new `slurmPolicy`, `lsfPolicy`, etc., for each new backend, leading to significant API bloat and violating the Open/Closed Principle.
```

Comment:
Will supporting a new workload managers require for us to update our extension framework with new plugin in any case? If yes, we still require to perform changes in the TrainJob controller.

Reply:
Oh, to be clear, my thinking was that HPCPolicy == covering a _set_ of managers (plugins). Does the mapping between policy and manager have to be 1:1? I was using the manager name as the determining variable for when a plugin should trigger. Technically, any/all of them can be called, but then if the manager name doesn't match the plugin we return early.

Reply:
Agree (discussion in slack) and resolving here. We are targeting the first example above with `flux` directly under the `mlPolicy`.